### PR TITLE
VPN-4064 - Add session timing metrics

### PR DESCRIPTION
**Pull request by mcleinman**
## Description

Adding metrics for session timing. From my testing (using the "log pings" debug option, this seems to be working well.

## Reference

https://mozilla-hub.atlassian.net/browse/VPN-4064

## Checklist
    
- [X] My code follows the style guidelines for this project
- [X] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [X] I have performed a self review of my own code
- [X] I have commented my code PARTICULARLY in hard to understand areas
- [X] I have added thorough tests where needed

### CODE REVIEW

**Comment by mcleinman** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

(2 of these metrics were [previously approved](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/5702#issuecomment-1397761161), so this form is for the one new one.)

1) What questions will you answer with this data?

These are a handful of metrics to learn more about how people use the VPN app. How long are users keeping a VPN session active for? What sort of VPN connectivity issues do users face, and how often? 

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? 

We do not have great insight into some of basic questions about our users: How often are they having issues? What sort of issues are they? How long do users keep the VPN active for?

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

Carefully considered which metrics were needed.

4) Can current instrumentation answer these questions?

No, unfortunately.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Name</td>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>sessions_started</td>
    <td>Count of sessions started in a given ping, to see how pings have edge cases in session starts/ends</td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-4064">VPN-4064</a></td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.

This collection is documented in the Glean Dictionary at https://dictionary.telemetry.mozilla.org/"

7) How long will this data be collected?  Choose one of the following:

* I want to permanently monitor this data. mcleinman@mozilla.com

8) What populations will you measure?

No filters - all channels, countries, and locales, unless the user has opted out of data collection on that device.

9) If this data collection is default on, what is the opt-out mechanism for users?

When launching the app for the first time, a user is given an option for whether to allow data collection. After this initial set up screen, a user can always toggle data collection permissions in the System Preferences screen.

10) Please provide a general description of how you will analyze this data.

Dashboards that product managers and others will consult on a regular basis.

11) Where do you intend to share the results of your analysis?

Within the Mozilla VPN team and Security and Privacy Products team.

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? 

No

**Comment by mcleinman** [comment]
Tagging @travis79 for data-review

**Review by bakulf — COMMENTED** [review]
Can we move this into telemetry.cpp?

**Comment by mcleinman** [comment]
> Can we move this into telemetry.cpp?

So I understand, is your hope that all glean calls are in the backend, all glean calls are in `telemetry.cpp`, or something else?

**Comment by bakulf** [comment]
Correct. Let's use signals and slots to unify the glean event generation. Instead of spread them in the code. Easy to maintain, easy to develop.

**Comment by mcleinman** [comment]
> Correct. Let's use signals and slots to unify the glean event generation. Instead of spread them in the code. Easy to maintain, easy to develop.

I'm not sure I 100% agree. I'm of the opinion that telemetry/metrics should be as lightweight and straightforward as possible, with minimal additional code. I think adding slots/signals where they don't exist just to move telemetry code to a separate file makes our code harder to follow, not easier - and adds more code than is needed for the task. (When following that code path, a developer needs to see where connects to that signal - and if it's in multiple spots - rather than just seeing a single line logging some metric.) While metrics/telemetry all over the codebase spreads them out, it's trivial to find all metrics with a global find.

If there is already a signal/slot setup for that task, it is less additional code to move a metric to the telemetry file - so it seems more reasonable. But the more I'm thinking about it, the less I like centralizing the metrics in one file. That said, I'm open to however we want to do it. Should we bring this up at the next engineering meeting for a quick group chat?

**Comment by brizental** [comment]
> I'm not sure I 100% agree. I'm of the opinion that telemetry/metrics should be as lightweight and straightforward as possible, with minimal additional code. (In my mind, they should be as lightweight as a log line in the code.) I think adding slots/signals where they don't exist just to move telemetry code to a separate file makes our code harder to follow, not easier - and adds more code than is needed for the task.

Late to the party, but I agree.

File: qtglean/metrics.yaml
Comment (brizental) [inline qtglean/metrics.yaml:73]: These two metrics are essentially meaningless when the sessions_started count is > 1. Please include this note in the description of both metrics. They should always be analysed together.

File: src/apps/vpn/controller.cpp
Comment (brizental) [inline src/apps/vpn/controller.cpp:574]: Why is this submit here? I would prefer to have the scheduling implementation on a separate PR.

**Review by brizental — CHANGES_REQUESTED** [review]
Please add unit tests for these metrics. Also, please include the Jira ticket on the PR title :)

**Comment by mcleinman** [comment]
> > I'm not sure I 100% agree. I'm of the opinion that telemetry/metrics should be as lightweight and straightforward as possible, with minimal additional code. (In my mind, they should be as lightweight as a log line in the code.) I think adding slots/signals where they don't exist just to move telemetry code to a separate file makes our code harder to follow, not easier - and adds more code than is needed for the task.
> 
> Late to the party, but I agree.

@bakulf Are you okay with this approach?

**Review by travis79 — APPROVED** [review]
> # Request for data collection review form
> 
> **All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**
> 
> (2 of these metrics were [previously approved](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/5702#issuecomment-1397761161), so this form is for the one new one.)
> 
>     1. What questions will you answer with this data?
> 
> 
> These are a handful of metrics to learn more about how people use the VPN app. How long are users keeping a VPN session active for? What sort of VPN connectivity issues do users face, and how often?
> 
>     2. Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?
> 
> 
> We do not have great insight into some of basic questions about our users: How often are they having issues? What sort of issues are they? How long do users keep the VPN active for?
> 
>     3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
> 
> 
> Carefully considered which metrics were needed.
> 
>     4. Can current instrumentation answer these questions?
> 
> 
> No, unfortunately.
> 
>     5. List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.
> 
> 
> **Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**
> Measurement Name 	Measurement Description 	Data Collection Category 	Tracking Bug #
> sessions_started 	Count of sessions started in a given ping, to see how pings have edge cases in session starts/ends 	interaction 	[VPN-4064](https://mozilla-hub.atlassian.net/browse/VPN-4064)
> 
>     6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
> 
> 
> This collection is documented in the Glean Dictionary at [dictionary.telemetry.mozilla.org](https://dictionary.telemetry.mozilla.org/)"
> 
>     7. How long will this data be collected?  Choose one of the following:
> 
> 
>     * I want to permanently monitor this data. [mcleinman@mozilla.com](mailto:mcleinman@mozilla.com)
> 
> 
>     8. What populations will you measure?
> 
> 
> No filters - all channels, countries, and locales, unless the user has opted out of data collection on that device.
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
> When launching the app for the first time, a user is given an option for whether to allow data collection. After this initial set up screen, a user can always toggle data collection permissions in the System Preferences screen.
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
> Dashboards that product managers and others will consult on a regular basis.
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
> Within the Mozilla VPN team and Security and Privacy Products team.
> 
>     12. Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection?
> 
> 
> No

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/).

2) Is there a control mechanism that allows the user to turn the data collection on and off?

Yes, through the data preferences in the application settings.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

[mcleinman@mozilla.com](mailto:mcleinman@mozilla.com) to be responsible for monitoring the permanent collection of this data

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 2, Interaction data

5) Is the data collection request for default-on or default-off?

Default-on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

No

7) Is the data collection covered by the existing Firefox privacy notice?

Yes

8) Does the data collection use a third-party collection tool?

No

## Result

data-review+

**Comment by mcleinman** [comment]
DO NOT REVIEW CODE, THIS IS NOW BLOCKED ON BRINGING UUID GLEAN METRIC IN

**Comment by mcleinman** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

(2 of these metrics were [previously approved](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/5702#issuecomment-1397761161), so this form is for the one new one.)

1) What questions will you answer with this data?

These are a handful of metrics to learn more about how people use the VPN app. How long are users keeping a VPN session active for? What sort of VPN connectivity issues do users face, and how often? 

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? 

We do not have great insight into some of basic questions about our users: How often are they having issues? What sort of issues are they? How long do users keep the VPN active for?

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

Carefully considered which metrics were needed.

4) Can current instrumentation answer these questions?

No, unfortunately.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Name</td>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>session_id</td>
    <td>A unique ID to identify each VPN session. This is used to connect a session across pings. (We need a user lifetime as we need it to persist beyond application runs, as VPN sessions last longer than app runs on mobile.) This ID will be created by the app and stored in Glean, and also be shared to the daemon Glean instance (which is coming in a separate data review).</td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-4064">VPN-4064</a></td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.

This collection is documented in the Glean Dictionary at https://dictionary.telemetry.mozilla.org/"

7) How long will this data be collected?  Choose one of the following:

* I want to permanently monitor this data. mcleinman@mozilla.com

8) What populations will you measure?

No filters - all channels, countries, and locales, unless the user has opted out of data collection on that device.

9) If this data collection is default on, what is the opt-out mechanism for users?

When launching the app for the first time, a user is given an option for whether to allow data collection. After this initial set up screen, a user can always toggle data collection permissions in the System Preferences screen.

10) Please provide a general description of how you will analyze this data.

Dashboards that product managers and others will consult on a regular basis.

11) Where do you intend to share the results of your analysis?

Within the Mozilla VPN team and Security and Privacy Products team.

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? 

No

**Comment by mcleinman** [comment]
Tagging @travis79 for an [additional data review](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/6303#issuecomment-1484249879), given that the data collection details changed last week.

**Review by travis79 — APPROVED** [review]
> # Request for data collection review form
> 
> **All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**
> 
> (2 of these metrics were [previously approved](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/5702#issuecomment-1397761161), so this form is for the one new one.)
> 
>     1. What questions will you answer with this data?
> 
> 
> These are a handful of metrics to learn more about how people use the VPN app. How long are users keeping a VPN session active for? What sort of VPN connectivity issues do users face, and how often?
> 
>     2. Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?
> 
> 
> We do not have great insight into some of basic questions about our users: How often are they having issues? What sort of issues are they? How long do users keep the VPN active for?
> 
>     3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
> 
> 
> Carefully considered which metrics were needed.
> 
>     4. Can current instrumentation answer these questions?
> 
> 
> No, unfortunately.
> 
>     5. List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.
> 
> 
> **Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**
> Measurement Name 	Measurement Description 	Data Collection Category 	Tracking Bug #
> session_id 	A unique ID to identify each VPN session. This is used to connect a session across pings. (We need a user lifetime as we need it to persist beyond application runs, as VPN sessions last longer than app runs on mobile.) This ID will be created by the app and stored in Glean, and also be shared to the daemon Glean instance (which is coming in a separate data review). 	interaction 	[VPN-4064](https://mozilla-hub.atlassian.net/browse/VPN-4064)
> 
>     6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
> 
> 
> This collection is documented in the Glean Dictionary at [dictionary.telemetry.mozilla.org](https://dictionary.telemetry.mozilla.org/)"
> 
>     7. How long will this data be collected?  Choose one of the following:
> 
> 
>     * I want to permanently monitor this data. [mcleinman@mozilla.com](mailto:mcleinman@mozilla.com)
> 
> 
>     8. What populations will you measure?
> 
> 
> No filters - all channels, countries, and locales, unless the user has opted out of data collection on that device.
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
> When launching the app for the first time, a user is given an option for whether to allow data collection. After this initial set up screen, a user can always toggle data collection permissions in the System Preferences screen.
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
> Dashboards that product managers and others will consult on a regular basis.
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
> Within the Mozilla VPN team and Security and Privacy Products team.
> 
>     12. Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection?
> 
> 
> No

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/).

2) Is there a control mechanism that allows the user to turn the data collection on and off?

Yes, on first install and through the data preferences in the application settings.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

Permanent collection to be monitored over time by  [mcleinman@mozilla.com](mailto:mcleinman@mozilla.com)

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 2, Interaction data

5) Is the data collection request for default-on or default-off?

Default-on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

No

7) Is the data collection covered by the existing Firefox privacy notice?

Yes

8) Does the data collection use a third-party collection tool?

No

## Result

data-review+

File: src/apps/vpn/controller.cpp
Comment (bakulf) [inline src/apps/vpn/controller.cpp:583]: Submission

File: src/apps/vpn/controller.cpp
Comment (bakulf) [inline src/apps/vpn/controller.cpp:588]: You don't use `sessionId`. Don't create it.

File: src/apps/vpn/controller.cpp
Comment (bakulf) [inline src/apps/vpn/controller.cpp:583]: Are you going to address this Todo in this PR?

File: src/apps/vpn/controller.cpp
Comment (bakulf) [inline src/apps/vpn/controller.cpp:531]: If you don't use `sessionId`, don't create it.

**Review by bakulf — CHANGES_REQUESTED** [review]

File: qtglean/metrics.yaml
Comment (brizental) [inline qtglean/metrics.yaml:97]: Which edge cases? Please add further explanation here, this is too vague.

File: src/apps/vpn/controller.cpp
Comment (brizental) [inline src/apps/vpn/controller.cpp:585]: One of the triggers is to send the ping right before the start of a new session. So this is false.

**Review by brizental — APPROVED** [review]
r+wc. Just some documentation nits.

File: src/apps/vpn/controller.cpp
Comment (mcleinman) [reply src/apps/vpn/controller.cpp:588]: We're going to use it in https://mozilla-hub.atlassian.net/browse/VPN-4038, coming shortly after this. Given that this is all building on each other, I have a slight preference towards leaving this variable for now. But if you feel strongly, I'll happily remove.

File: src/apps/vpn/controller.cpp
Comment (mcleinman) [reply src/apps/vpn/controller.cpp:531]: Same as above, it will be used in VPN-4038.

File: src/apps/vpn/controller.cpp
Comment (mcleinman) [reply src/apps/vpn/controller.cpp]: It's a note for https://mozilla-hub.atlassian.net/browse/VPN-4443. I updated the wording to remove the TODO, and make it clear it's a note for when working that ticket.

Comment (mcleinman) [reply src/apps/vpn/controller.cpp]: Thanks, updated.

File: src/apps/vpn/controller.cpp
Comment (mcleinman) [reply src/apps/vpn/controller.cpp:586]: Not according to the tech doc that this is based on: https://mozilla-hub.atlassian.net/wiki/spaces/SECPRV/pages/184419319/Metric+pings+-+Tech+spec#Session-ping

If you have a problem with that doc, let's discuss it on the doc.

Comment (brizental) [reply src/apps/vpn/controller.cpp:586]: Aha! Yes I have a problem with that indeed! Thanks for flagging it Matt, it was an oversight on my part not to comment on my first pass.

Comment (mcleinman) [reply src/apps/vpn/controller.cpp]: Glad you noticed this! I've updated the comment here to reflect our updated plan for ping submission.

**Comment by mcleinman** [comment]
@bakulf Let me know if you want to do another review on this.

**Pull request merged**
